### ISSUE TITLE
Reconnect on connection closed

### ISSUE THREAD

**Issue by james-pellow** [theophilusx/ssh2-sftp-client#217]
We are building an upload tool that must be resilient to connection loss.  Right now when the connection is lost while an append or put operation is in progress, we get a "close" event, but the append or put operations don't resolve or throw.  Currently I'm wrapping our upload as such:

```
    try {
      const success = await new Promise<boolean>(resolv => {
        const listener = (s: boolean) => {
          this.client.removeListener("close", listener);
          resolv(s);
        };
        this.client.on("close", () => listener(false));

        // If the connection is lost, neither of these ever resolve.
        if (tryCont) {
          this.client.append(stream, destFile).then(() => listener(true));
        } else {
          this.client.put(stream, destFile).then(() => listener(true));
        }
      });

      if (!success) throw new Error("Connection lost");

    } catch (e) {
      console.log(e);
      throw e;
    }
```

This works, but feels like its leaking in the append and put calls since they are not resolving.

Along the same lines, calling connect multiple times to test for a connection results in the sftp server erroring out with a long list of "no more sessons" errors when a connection is finally made.  Its like each failed connect call finally makes a connection even when it reports failure when its initially called.  Is there an intended approach for doing this kind of thing?  Thanks!

**Comment by theophilusx** [comment]
This is quite odd. What version of ssh2-sftp-client are you using and on what version of node and what platform?

How are you 'creating' the lost connections? When a connection is lost unexpectedly, ssh2 should be raising an error signal. Both put and append listen for that error and will reject the promise if one is raised.  Note that a close signal is raised in 2 situations - when an error occurs and the connection is closed and when a end/close is requested (no error). If you are seeing a close() signal, but no error signal, then that indicates the connection is being closed after a request, which is not  an error (and probably why the put/append do not reject.  I guess we could change things to always reject if a close signal is raised. My only concern is this will result in errors being reported when technically, it isn't an error. I was a bit reluctant to do this as it would likely result in more issues being logged (by far, logged issues are due to user code error rather than module bugs). Might need to reconsider that.

With regards to making connections. You need to be very careful about trying to connect too many times in a short period. There is considerable overhead in establishing ssh connections. During the 4establishment of a connection, there is a period where the remote server sees the connection as an 'unauthenticated' connection (prior to handshake, key exchange, cipher negotiation etc). Most sftp servers have a form of rate limiting where they will only allow a small (i.e. 10) number of unauthenticated connections. After this initial limit is reached, it will begin to randomly 'drop' a percentage of connections until a hard limit (i.e. 15) is reached and all connection attempts are dropped. The result is what can appear as random connection failures, followed by complete connection failures.  Note that ssh2-sftp-client does include a retry mechanism for connections. It will attempt to connect multiple times (I think it is 2 or 3 by default - you can change this) before giving up. Perhaps you are tripping over this? How are you determining that the connection has failed (waiting for promise rejection or using signals?)

**Comment by james-pellow** [comment]
Thanks for your quick response!  First, I'm using ssh2-sftp-client version 5.1.0 on node 12.13.0 in electron.js on osx 10.15.3.  Here is my upload service code:

```
import FtpClient from "ssh2-sftp-client";
import { EventEmitter } from "events";
import { Readable } from "stream";

export declare interface FtpUpload {
  on(event: "progress", listener: (percent: number) => void): this;
  on(event: "started" | "finished", listener: () => void): this;
  on(event: "error", listener: (description: string) => void): this;
}

export class FtpUpload extends EventEmitter {
  private client = new FtpClient();
  private connected = false;
  private progressTimer = -1;
  private progress = 0;
  private isCanceling = false;

  constructor() {
    super();
  }

  public get isConnected() {
    return this.connected;
  }

  public async connect(
    host: string,
    port: number,
    username: string,
    password: string,
  ) {
    if (this.connected) return;
    try {
      await this.client.connect({
        host,
        port,
        username,
        password,
        retries: 1000,
        retry_factor: 1,
        retry_minTimeout: 5000,
      } as any);
      this.connected = true;
    } catch (e) {
      console.log("Failed to connect:", e);
      this.connected = false;
    }
  }

  public async download(file: string): Promise<string | undefined> {
    if (!this.connected) return undefined;
    try {
      const data = (await this.client.get(file)) as Buffer;
      return data.toString();
    } catch (e) {
      console.log("Failed to retrieve sftp download: ", e);
      return undefined;
    }
  }

  /**
   * Upload a specified buffer to the remote ftp destination
   *
   * @param srcBuff Source buffer
   * @param destFile Destination path
   * @param tryCont Try to resume file
   * @returns True if successful, false if canceled
   */
  public async upload(
    srcBuff: Buffer,
    destFile: string,
    tryCont: boolean,
  ): Promise<boolean> {
    if (!this.connected) return true;

    let filePos = 0;
    if (tryCont) {
      const destSize = await this.getDestFileSize(destFile);
      if (destSize !== undefined) {
        if (destSize === srcBuff.length) return true;
        this.emit("progress", 100);
        filePos = destSize;
      } else {
        tryCont = false;
      }
    }

    const buff = srcBuff.slice(filePos);
    const startPercent = (100 * filePos) / srcBuff.length;

    this.emit("started");
    this.emit("progress", startPercent);
    this.progress = 0;
    this.progressTimer = window.setInterval(
      () => this.emit("progress", this.progress),
      1000,
    );

    // We can't use the path system because it won't properly handle
    // paths if we are on windows.
    let path = destFile;
    const pos = destFile.lastIndexOf("/");
    if (pos >= 0) path = destFile.slice(0, pos);
    console.log("Creating path:", path);
    await this.client.mkdir(path, true);

    const sliceSize = buff.length / (100 - startPercent);
    let start = 0;

    const stream = new Readable({
      read: async () => {
        // WARNING: For testing only! Don't leave un-commented when deployed!
        await new Promise(r => setTimeout(r, 1000));
        if (this.isCanceling) {
          stream.push(null);
          return;
        }
        const end = Math.min(start + sliceSize, buff.length);
        stream.push(buff.slice(start, end));
        start = end;

        if (start >= buff.length) stream.push(null);
      },
    });

    let total = filePos;
    stream.on("data", b => {
      total += b.length;
      const progress = (100 * total) / srcBuff.length;
      this.progress = progress;
    });

    try {
      const success = await new Promise<boolean>(resolv => {
        const listener = (s: boolean) => {
          if (!s) this.connected = false;
          this.client.removeListener("close", listener);
          resolv(s);
        };
        this.client.on("close", () => listener(false));

        // If the connection is lost, neither of these ever resolve.
        if (tryCont) {
          this.client.append(stream, destFile).then(() => listener(true));
        } else {
          this.client.put(stream, destFile).then(() => listener(true));
        }
      });

      if (!success) throw new Error("Connection lost");

      if (this.isCanceling) {
        await this.client.delete(destFile);
        this.client.end();
        this.connected = false;
      }
    } catch (e) {
      console.log(e);
      this.clearProgressTimer();
      throw e;
    }

    this.clearProgressTimer();
    this.emit("progress", this.progress);
    this.emit("finished");
    return !this.isCanceling;
  }

  public cancel() {
    this.isCanceling = true;
  }

  private clearProgressTimer(): void {
    if (this.progressTimer) window.clearInterval(this.progressTimer);
  }

  private async getDestFileSize(name: string): Promise<number | undefined> {
    if (!this.connected) {
      console.error("Attempting to get file size without connection");
      return undefined;
    }

    try {
      const stats = await this.client.stat(name);
      return stats.size;
    } catch {
      console.log("File not found in destination", name);
      return undefined;
    }
  }
}
```
I implemented your suggestion of using connect's retry logic.  As you can see, it reconnects every 5 seconds.  If we can't connect after 1000 retries in connect, I restart the process after another 5 seconds.  That is handled outside of this class.  Outages during an upload are detected by the upload method throwing an exception.  If the upload method throws, I retry the upload every 5 seconds.  That retry logic includes calling connect if we are not already connected.  I was aware of the ssh connection rate limiting logic, but I'm pretty sure I'm not getting close to that in my code.  Now for the issue.  If I allow the connection to be made, and get an upload started (15:50:02.954), then part way through the upload I kill the server(15:50:19.944), then I let it retry to connect for a while(15:50:29.924 - 15:50:49.962), then I start the server, and let it connect(15:50:56.006), then I kill the server again(15:51:05.455). I get multiple console.errors stating "sftp: Connection closed unexpectedly" apparently two for each failed connection attempt from before.  Then, each reconnect appears to retry that same number of times for each retry, as though there are a bunch of clients running in parallel in the background.  Here is my console log:

```
15:49:26.357 log.ts:24 [HMR] Waiting for update signal from WDS...
15:50:02.787 upload-queue.ts:187 Uploading /Users/pellja/Projects/AWR/uploader-app/examples/ORMAD_AWRx_20200215_7.mp3 to /AD/MP/ORMZL/unchecked/AWR/ORMAD_AWRx_2020217_3.mp3
15:50:02.923 upload-queue.ts:221 Starting upload
15:50:02.954 ftpupload.service.ts:104 Creating path: /upload/AD/MP/ORMZL/unchecked/AWR
15:50:19.944 index.ts:1 sftp: Connection closed unexpectedly
15:50:19.945 index.ts:1 sftp: Connection closed unexpectedly
15:50:19.946 ftpupload.service.ts:158 Error: Connection lost
15:50:19.946 upload-queue.ts:230 Upload failed, retrying later... Error: Connection lost
15:50:29.924 index.ts:1 Error event: connect ECONNREFUSED 127.0.0.1:2222
15:50:29.924 index.ts:1 sftp: Connection closed unexpectedly
15:50:34.951 index.ts:1 Error event: connect ECONNREFUSED 127.0.0.1:2222
15:50:34.951 index.ts:1 sftp: Connection closed unexpectedly
15:50:39.954 index.ts:1 Error event: connect ECONNREFUSED 127.0.0.1:2222
15:50:39.954 index.ts:1 sftp: Connection closed unexpectedly
15:50:44.959 index.ts:1 Error event: connect ECONNREFUSED 127.0.0.1:2222
15:50:44.959 index.ts:1 sftp: Connection closed unexpectedly
15:50:49.962 index.ts:1 Error event: connect ECONNREFUSED 127.0.0.1:2222
15:50:49.962 index.ts:1 sftp: Connection closed unexpectedly
15:50:56.006 upload-queue.ts:187 Uploading /Users/pellja/Projects/AWR/uploader-app/examples/ORMAD_AWRx_20200215_7.mp3 to /AD/MP/ORMZL/unchecked/AWR/ORMAD_AWRx_2020217_3.mp3
15:50:56.132 upload-queue.ts:221 Starting upload
15:50:56.164 ftpupload.service.ts:104 Creating path: /upload/AD/MP/ORMZL/unchecked/AWR
15:51:05.455 index.ts:1 sftp: Connection closed unexpectedly
15:51:05.455 index.ts:1 sftp: Connection closed unexpectedly
15:51:05.455 index.ts:1 sftp: Connection closed unexpectedly
15:51:05.455 index.ts:1 sftp: Connection closed unexpectedly
15:51:05.455 index.ts:1 sftp: Connection closed unexpectedly
15:51:05.455 index.ts:1 sftp: Connection closed unexpectedly
15:51:05.456 index.ts:1 sftp: Connection closed unexpectedly
15:51:05.456 index.ts:1 sftp: Connection closed unexpectedly
15:51:05.457 ftpupload.service.ts:158 Error: Connection lost
15:51:05.479 upload-queue.ts:230 Upload failed, retrying later... Error: Connection lost
15:51:10.543 index.ts:1 Error event: connect ECONNREFUSED 127.0.0.1:2222
15:51:10.543 index.ts:1 Error event: connect ECONNREFUSED 127.0.0.1:2222
15:51:10.543 index.ts:1 Error event: connect ECONNREFUSED 127.0.0.1:2222
15:51:10.543 index.ts:1 Error event: connect ECONNREFUSED 127.0.0.1:2222
15:51:10.544 index.ts:1 Error event: connect ECONNREFUSED 127.0.0.1:2222
15:51:10.544 index.ts:1 Error event: connect ECONNREFUSED 127.0.0.1:2222
15:51:10.544 index.ts:1 Error event: connect ECONNREFUSED 127.0.0.1:2222
15:51:10.544 index.ts:1 Error event: connect ECONNREFUSED 127.0.0.1:2222
15:51:10.544 index.ts:1 sftp: Connection closed unexpectedly
15:51:10.544 index.ts:1 sftp: Connection closed unexpectedly
15:51:10.544 index.ts:1 sftp: Connection closed unexpectedly
15:51:10.545 index.ts:1 sftp: Connection closed unexpectedly
15:51:10.545 index.ts:1 sftp: Connection closed unexpectedly
15:51:10.545 index.ts:1 sftp: Connection closed unexpectedly
15:51:10.545 index.ts:1 sftp: Connection closed unexpectedly
15:51:10.545 index.ts:1 sftp: Connection closed unexpectedly
15:51:15.570 index.ts:1 Error event: connect ECONNREFUSED 127.0.0.1:2222
15:51:15.571 index.ts:1 Error event: connect ECONNREFUSED 127.0.0.1:2222
15:51:15.571 index.ts:1 Error event: connect ECONNREFUSED 127.0.0.1:2222
15:51:15.571 index.ts:1 Error event: connect ECONNREFUSED 127.0.0.1:2222
15:51:15.571 index.ts:1 Error event: connect ECONNREFUSED 127.0.0.1:2222
15:51:15.571 index.ts:1 Error event: connect ECONNREFUSED 127.0.0.1:2222
15:51:15.571 index.ts:1 Error event: connect ECONNREFUSED 127.0.0.1:2222
15:51:15.571 index.ts:1 Error event: connect ECONNREFUSED 127.0.0.1:2222
15:51:15.571 index.ts:1 sftp: Connection closed unexpectedly
15:51:15.572 index.ts:1 sftp: Connection closed unexpectedly
15:51:15.572 index.ts:1 sftp: Connection closed unexpectedly
15:51:15.572 index.ts:1 sftp: Connection closed unexpectedly
15:51:15.572 index.ts:1 sftp: Connection closed unexpectedly
15:51:15.572 index.ts:1 sftp: Connection closed unexpectedly
15:51:15.572 index.ts:1 sftp: Connection closed unexpectedly
15:51:15.572 index.ts:1 sftp: Connection closed unexpectedly
```
Now if I let it go like that for a while, retrying the connection, then I start the server and allow it to reconnect, the server seems to connect and then I get hundreds of "sftp: Connection closed unexpectedly" and the server logs:

```
ftp_1  | no more sessions
ftp_1  | no more sessions
ftp_1  | no more sessions
ftp_1  | no more sessions
ftp_1  | no more sessions
ftp_1  | no more sessions
ftp_1  | no more sessions
ftp_1  | no more sessions
ftp_1  | no more sessions
ftp_1  | no more sessions
ftp_1  | no more sessions
ftp_1  | no more sessions
ftp_1  | no more sessions
ftp_1  | no more sessions
ftp_1  | no more sessions
ftp_1  | no more sessions
ftp_1  | no more sessions
ftp_1  | no more sessions
ftp_1  | no more sessions
ftp_1  | no more sessions
ftp_1  | no more sessions
ftp_1  | no more sessions
ftp_1  | no more sessions
ftp_1  | no more sessions
ftp_1  | no more sessions
ftp_1  | no more sessions
ftp_1  | no more sessions
ftp_1  | no more sessions
ftp_1  | no more sessions
ftp_1  | no more sessions
ftp_1  | no more sessions
ftp_1  | no more sessions
ftp_1  | no more sessions
ftp_1  | no more sessions
ftp_1  | no more sessions
ftp_1  | no more sessions
ftp_1  | no more sessions
ftp_1  | no more sessions
ftp_1  | no more sessions
ftp_1  | no more sessions
ftp_1  | no more sessions
ftp_1  | no more sessions
ftp_1  | no more sessions
ftp_1  | no more sessions
ftp_1  | Received disconnect from 172.18.0.1 port 59608:11:
ftp_1  | Disconnected from 172.18.0.1 port 59608
ftp_1  | Received disconnect from 172.18.0.1 port 59612:11:  [preauth]
ftp_1  | Disconnected from 172.18.0.1 port 59612 [preauth]
ftp_1  | Received disconnect from 172.18.0.1 port 59616:11:  [preauth]
ftp_1  | Disconnected from 172.18.0.1 port 59616 [preauth]
ftp_1  | Received disconnect from 172.18.0.1 port 59620:11:  [preauth]
ftp_1  | Disconnected from 172.18.0.1 port 59620 [preauth]
ftp_1  | Received disconnect from 172.18.0.1 port 59624:11:  [preauth]
ftp_1  | Disconnected from 172.18.0.1 port 59624 [preauth]
ftp_1  | Received disconnect from 172.18.0.1 port 59628:11:  [preauth]
ftp_1  | Disconnected from 172.18.0.1 port 59628 [preauth]
ftp_1  | Received disconnect from 172.18.0.1 port 59632:11:  [preauth]
ftp_1  | Disconnected from 172.18.0.1 port 59632 [preauth]
ftp_1  | Received disconnect from 172.18.0.1 port 59636:11:  [preauth]
ftp_1  | Disconnected from 172.18.0.1 port 59636 [preauth]
ftp_1  | Received disconnect from 172.18.0.1 port 59640:11:  [preauth]
ftp_1  | Disconnected from 172.18.0.1 port 59640 [preauth]
ftp_1  | Received disconnect from 172.18.0.1 port 59644:11:  [preauth]
ftp_1  | Disconnected from 172.18.0.1 port 59644 [preauth]
ftp_1  | Received disconnect from 172.18.0.1 port 59648:11:  [preauth]
ftp_1  | Disconnected from 172.18.0.1 port 59648 [preauth]
ftp_1  | Received disconnect from 172.18.0.1 port 59652:11:  [preauth]
ftp_1  | Disconnected from 172.18.0.1 port 59652 [preauth]
ftp_1  | Received disconnect from 172.18.0.1 port 59656:11:  [preauth]
ftp_1  | Disconnected from 172.18.0.1 port 59656 [preauth]
ftp_1  | Received disconnect from 172.18.0.1 port 59660:11:  [preauth]
ftp_1  | Disconnected from 172.18.0.1 port 59660 [preauth]
ftp_1  | Received disconnect from 172.18.0.1 port 59664:11:  [preauth]
ftp_1  | Disconnected from 172.18.0.1 port 59664 [preauth]
ftp_1  | Received disconnect from 172.18.0.1 port 59668:11:  [preauth]
ftp_1  | Disconnected from 172.18.0.1 port 59668 [preauth]
ftp_1  | Received disconnect from 172.18.0.1 port 59672:11:  [preauth]
ftp_1  | Disconnected from 172.18.0.1 port 59672 [preauth]
ftp_1  | Received disconnect from 172.18.0.1 port 59676:11:  [preauth]
ftp_1  | Disconnected from 172.18.0.1 port 59676 [preauth]
ftp_1  | Received disconnect from 172.18.0.1 port 59680:11:  [preauth]
ftp_1  | Disconnected from 172.18.0.1 port 59680 [preauth]
ftp_1  | Received disconnect from 172.18.0.1 port 59684:11:  [preauth]
ftp_1  | Disconnected from 172.18.0.1 port 59684 [preauth]
ftp_1  | Received disconnect from 172.18.0.1 port 59688:11:  [preauth]
ftp_1  | Disconnected from 172.18.0.1 port 59688 [preauth]
ftp_1  | Received disconnect from 172.18.0.1 port 59692:11:  [preauth]
ftp_1  | Disconnected from 172.18.0.1 port 59692 [preauth]
ftp_1  | Received disconnect from 172.18.0.1 port 59696:11:  [preauth]
ftp_1  | Disconnected from 172.18.0.1 port 59696 [preauth]
ftp_1  | Received disconnect from 172.18.0.1 port 59700:11:  [preauth]
ftp_1  | Disconnected from 172.18.0.1 port 59700 [preauth]
ftp_1  | Received disconnect from 172.18.0.1 port 59704:11:  [preauth]
ftp_1  | Disconnected from 172.18.0.1 port 59704 [preauth]
ftp_1  | Received disconnect from 172.18.0.1 port 59708:11:  [preauth]
ftp_1  | Disconnected from 172.18.0.1 port 59708 [preauth]
ftp_1  | Received disconnect from 172.18.0.1 port 59712:11:  [preauth]
ftp_1  | Disconnected from 172.18.0.1 port 59712 [preauth]
ftp_1  | Received disconnect from 172.18.0.1 port 59716:11:  [preauth]
ftp_1  | Disconnected from 172.18.0.1 port 59716 [preauth]
ftp_1  | Received disconnect from 172.18.0.1 port 59720:11:  [preauth]
ftp_1  | Disconnected from 172.18.0.1 port 59720 [preauth]
ftp_1  | Received disconnect from 172.18.0.1 port 59724:11:  [preauth]
ftp_1  | Disconnected from 172.18.0.1 port 59724 [preauth]
ftp_1  | Received disconnect from 172.18.0.1 port 59728:11:  [preauth]
ftp_1  | Disconnected from 172.18.0.1 port 59728 [preauth]
ftp_1  | Received disconnect from 172.18.0.1 port 59732:11:  [preauth]
ftp_1  | Disconnected from 172.18.0.1 port 59732 [preauth]
ftp_1  | Received disconnect from 172.18.0.1 port 59736:11:  [preauth]
ftp_1  | Disconnected from 172.18.0.1 port 59736 [preauth]
ftp_1  | Received disconnect from 172.18.0.1 port 59740:11:  [preauth]
ftp_1  | Disconnected from 172.18.0.1 port 59740 [preauth]
ftp_1  | Received disconnect from 172.18.0.1 port 59744:11:  [preauth]
ftp_1  | Disconnected from 172.18.0.1 port 59744 [preauth]
ftp_1  | Received disconnect from 172.18.0.1 port 59748:11:  [preauth]
ftp_1  | Disconnected from 172.18.0.1 port 59748 [preauth]
ftp_1  | Received disconnect from 172.18.0.1 port 59752:11:  [preauth]
ftp_1  | Disconnected from 172.18.0.1 port 59752 [preauth]
ftp_1  | Received disconnect from 172.18.0.1 port 59756:11:  [preauth]
ftp_1  | Disconnected from 172.18.0.1 port 59756 [preauth]
... truncated
```
Its like it is trying to connect hundreds of times all at once and the rate limiting mechanism is kicking in.  I've verified that I'm not calling connect hundreds of times.  This appears to be happening in the library itself.  Any thoughts?  Thanks for your help!

**Comment by james-pellow** [comment]
Also, in response to your comment 

> ﻿﻿When a connection is lost unexpectedly, ssh2 should be raising an error signal. Both put and append listen for that error and will reject the promise if one is raised.

That doesn't seem to be happening.  If I kill the server, the promise is not rejected.  The only way I can catch that situation is to handle the on("close",) event.  I'm simulating a lost internet connection by closing the server.  I know that isn't an entirely accurate simulation, but its a start and should be handled as well.

**Comment by theophilusx** [comment]
There is a bit for me to digest there, so may take a bit. I also see
that your using typscript, which I'm not familiar with (been on the todo
list for a while!).

The built-in retry mechanism only attempts to connect 2 or 3 times (I
cannot remember - need to look at the code) and it has an increased
timeout between each attempt. After that it just stops trying. So, there is
nothing in the code I'm aware of which would result in large numbers of
connect attempts. It is pretty straight-forward - have a look at the
connect() method. You can also set the number of retries via the connect
config object. You could try setting that to 1 and setting it to a
larger number and see what effect that has.

the following is just some thoughts (stream of consciousness) and may
not be correct - just thinking out loud .....

One thing which may be happening that is complicating matters is that
the way you are closing down the server may be causing the ftp server to
hard exit before it can send any error signal. One thing you could try
is to add an on error handler and log it. If your not seeing any errors
signals, that would explain why the put/append methods are not being
rejected. If this is the case, then I guess I could add a close listener
to each method which will reject the promise. It should not be necessary
in 'normal' operation as the server should send an error signal if
something goes wrong and it needs to close the connection without a
request to do so (though, thinking about it, I guess a network failure
would still be a problem). One downside of doing this is that it may
result in duplicate or 'false positive' errors and I'll need to put
something in place to make sure 'normal' close events are not flagged as
errors (there is always a close event - even after an end() is called).
Technically, the only way to identify 'abnormal' (unexpected) and normal
(expected) close events is abnormal close events are proceeded by an
error event.

If I was to create a new branch in the repository and add a close
handler to each of the methods, would you be in a position to test that
version to see if it helps?

**Comment by theophilusx** [comment]
I just noticed that in your code, you have retries set to 1000. this would be what is causing all the connect attempts. A value of 1000 is excessive.

**Comment by iordachemc** [comment]
I also experience a problem of the same nature.
I'm connecting to a Kafka server and on 'message' I want to save a file to a SFTP server.
All works great until there is a problem with the connection to the SFTP server. To simulate that I am also abruptly forcefully closing the server. As a result I do not receive any error and the code hangs. I rare cases I do receive an error and I can restart the process and recover.

**Comment by theophilusx** [comment]
I have made some changes in the branch close-handler. Essentially, adding a close handler to each method which should result in the promise being rejected if a close signal is raised before it has completed/resolved. This should fix the issue of methods appearing to 'hang' when the connection is lost. Note however, if the remote server does not raise either an error signal or a close signal, hangs would still be possible. There isn't much I can do in that situation - if the client doens't recieve any signal to indicate the connection has closed or an error has occured, it can't do much. 

Anyway, please give the close-handler branch a go and see if that improves the istuation. If it does and there are no other unexpected side effects, I will release a new version of the module.

**Comment by james-pellow** [comment]
Thanks for your thoughtful reply and willingness to engage on this issue.  I sent a lot of information yesterday, and I appreciate your thoguhts.

I know you haven't fully digested my code, but I am using the retry mechanism built into connect.  I set the factor to 1 so that the retry rate doesn't decrease over time, and I have a fixed 5 second retry rate.  I set the retry count to 1000. That mechanism seems to work well initially, it just causes connection errors later.

I don't know if the server is sending error messages when it closes, however I suspect it is as the client notices immediately when I close the server.  But, as you noted, even if it doesn't that would just more accurately simulate a hard network failure such as a sudden disconnect, and I need to be able to handle that as well.

So I tested what events are received in the library when the server gets shut down.  We don't get an "error" event, but we do get a "close" event.  It seems like getting a close event during an active sftp operation would be an error, as the operation could not complete normally.  If we get close with no active operation then that could be just normal, so not even a change in behavior.  Just rejecting the sftp operation promises, as you suggest, would seem to be the logical behavior in the case of either error or close.  I wouldn't think of getting a rejected promise and also a close or error event to be a duplicate event as the rejected promise is just the logical outcome of the sftp operation ending unsuccessfully.  That would be my preference.

The second issue I'm having with the automatic connection retry mechanism is still a blocker for me.  Every time the connect is attempted, something is being hung onto and replayed to the server once a connection is made, as can be seen from the console log above.  What are your thoughts on that?  Thanks again!

**Comment by james-pellow** [comment]
I just noticed your comment earlier this morning.  I'll try your branch.  That sounds perfect!  Thanks.  That just leaves the reconnect issue, provided your new branch fixes the close and error cases.

**Comment by james-pellow** [comment]
I just tried the `close-handler` branch.  It almost works.  The close handler gets called twice when I close the server (not sure why twice...) and neither time is the reject method defined, so it just console.errors.  I don't see an obvious reason for that looking at the code.  I'm still investigating.

**Comment by james-pellow** [comment]
I appears the error and close listeners are being removed immediately on line 807 and 808 of index.js.  The try will finally out immediately even though the promise hasn't resolved.

**Comment by theophilusx** [comment]
Good catch. Bit of a brain fart there. I thought using finally to remove
the listeners would be clean, but overlooked the fact that the code
inside the try is async and therefore, not completed before the finally
is called.

have updated the code again and got rid of the finally blocks. Should
now only remove the listeners after the promise has been
resolved/rejected.

It is important to remove the listeners. If we don't, you will get the
'possible memory leak' error as every time you call a method, it would
just add new listeners and we will eventually exceed the default 10
limit.

Please try the code again.

**Comment by james-pellow** [comment]
That's an easy mistake to make.  I've done it a few times myself.  The new code is working perfectly!  Thank you!  Now, the only issue I'm facing is the connection issue described before.  Any ideas on that?  I'll poke at the code a bit as well.  Did I describe the issue before clearly?  I know I threw a lot of data at you at once there!

**Comment by james-pellow** [comment]
Looking at the code, it appears to me that there should be a return after line 111 in index.js.  If there is an error, we shouldn't continue on after the callback is executed.  Does that look right to you?

**Comment by james-pellow** [comment]
Hey!  I think I fixed it.  I'll get a PR for you shortly.  Probably Monday.  I just didn't want us both to be duplicating effort.  Have a great weekend!

**Comment by theophilusx** [comment]
Yes, I think your right. Once we have exhausted all attemptw and done
the error callback, we should just return. Have added it.

**Comment by iordachemc** [comment]
Thank you! 
I've tested and this has also solved my problem.

**Comment by james-pellow** [comment]
I've submitted a pull request ( #218 ) that solves the remaining issues on connect.  It was leaving ready event handlers was the primary issue, though it was also leaving behind close and error handlers.  Its clean at this point.  Thanks for all your help!

**Comment by theophilusx** [comment]
Hi james,

question regarding your PR.

You have added the following

    this.client.on('close', utils.makeCloseListener(this));
    this.client.on('error', err => {
      if (!this.errorHandled) {
        // error not already handled. Log it.
        console.error(`Error event: ${err.message}`);
      }
      //need to set to false in case another error raised
      this.errorHandled = false;
    });

Is this really needed? All that it does is send messages to the console
and when running my tests, it really just generates console noise. The
tests all pass (including those meant to fail) without it and I'm not
sure why it is needed.

**Comment by james-pellow** [comment]
That's just code I copied from the connect method. If you don't want it, I don't :)  Good call.

**Comment by theophilusx** [comment]
Issue resolved. Released in 5.1.1

**Comment by james-pellow** [comment]
Thanks!  I'll check it out!
